src/java/org/apache/cassandra/index/sai/disk/v1/DocLengthsReader.java

```diff
@@ -0,0 +1,59 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.cassandra.index.sai.disk.v1;
+
+import java.io.Closeable;
+import java.io.IOException;
+
+import org.apache.cassandra.index.sai.disk.io.IndexInputReader;
+import org.apache.cassandra.index.sai.utils.IndexFileUtils;
+import org.apache.cassandra.index.sai.utils.SAICodecUtils;
+import org.apache.cassandra.io.util.FileHandle;
+import org.apache.cassandra.io.util.FileUtils;
+import org.apache.lucene.codecs.CodecUtil;
+
+public class DocLengthsReader implements Closeable
+{
+    private final FileHandle fileHandle;
+    private final IndexInputReader input;
+    private final SegmentMetadata.ComponentMetadata componentMetadata;
+
+    public DocLengthsReader(FileHandle fileHandle, SegmentMetadata.ComponentMetadata componentMetadata)
+    {
+        this.fileHandle = fileHandle;
+        this.input = IndexFileUtils.instance.openInput(fileHandle);
+        this.componentMetadata = componentMetadata;
+    }
+
+    public int get(int rowID) throws IOException
+    {
+        // Account for header size in offset calculation
+        long position = componentMetadata.offset + (long) rowID * Integer.BYTES;
```

Comment:
I see this is called with an sstable row id, but the component metadata is per segment, which would imply it might expect a segment row id. Have you tested with segments? I am trying to get my env set up to run this test, but tests are failing with an unrelated error.

Reply:
In a test, this can work by setting:

```java
        SegmentBuilder.updateLastValidSegmentRowId(2);
        compact();
```

(Running compaction triggers the segmentation)

Reply:
I didn't get a test to fail, but I looked at the writer and found that we use the segment id when putting the doc length into the map: `docLengths.put(segmentRowId, terms.size());`

Reply:
We probably need better automated test coverage for randomly segmented indexes.

Reply:
I think we're good on this after the commit based on the other thread, code looks like this now
```
                    int docLength = docLengthsReader.get(rowId); // segment-local rowid
                    var pk = pkm.primaryKeyFromRowId(segmentRowIdOffset + rowId); // sstable-global rowid
                    return new DocTF(pk, docLength, merged.frequencies());
```

so DLR -- expects segment rowid, gets segment rowid
PKM -- expects sstable rowid, gets sstable rowid